### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by i1group**
**Overview**
This PR fixes:
1. `text` property was empty, because it's value was always set to `false` in constructor.
2. `getOptionsFromTag()` method was added to make it possible to initialize donutty using `data-` attributes any time / event / callback (not on document.ready only).

**Technical details**
1. This line in constructor
`this.options.text = false;`
was replaced with this
`this.options.text = isDefined( this.options.text ) ? this.options.text : false;`
2. Now we check if `this.$text` exists before set styles to it, as there were errors on empty text.
3. To initialize donutty using `data-` attribute params now you can call constructor with `null` value for options:
`new Donutty( el, null );`
4. Now donutty instance is always saved to `el` (container) and is accessible from outside like this
`$('.donutty-wrapper').get(0).donutty.setState(newState)`
There was an issue before: instance was saved to `el` only on document.ready and only if donutty was initialized using `data-` attributes (not working for js calls like `new Donutty(...)` at all).

### CODE REVIEW

**Comment by simeydotme** [comment]
Hey, thanks so much @i1group !
Overall this is looking good, and thanks so much for the clear codepens 🙇 

----

I think there's one issue though;

Although I like what you've done for initialising/auto-populating the options on a "vanilla" implementation, I think the idea has broken the jQuery api/pattern. With your proposed change, the user would no longer be able to run; `$(".my-el").donutty({ thickness: 3 });` without adding `data-` attributes to the element.

----

Let me see if I can make a few suggestions using the code review tool 🙂

File: src/donutty.js [hunk 1]
```diff
@@ -33,6 +33,18 @@
 
         }
 
+        if ( !this.$wrapper ) {
+
+            return this;
+
+        }
+
+        if (options === null) {
```
Comment (simeydotme) [inline src/donutty.js]:
```suggestion
        if ( typeof options === "undefined" ) {
```

File: src/donutty.js [hunk 2]
```diff
@@ -33,6 +33,18 @@
...
+
+            options = this.getOptionsFromTag();
```
Comment (simeydotme) [inline src/donutty.js]:
```suggestion
            options = JSON.parse( JSON.stringify( this.$wrapper.dataset ) );
```
Have to move this, as the declaration for `this.$wrapper` previously ran in `init()` which would not be available at runtime

File: src/donutty.js [hunk 3]
```diff
@@ -46,16 +58,24 @@
...
-        this.options.text =        false;
+        this.options.text =        isDefined( this.options.text ) ? this.options.text : false;
 
         this.init();
 
         return this;
 
     };
 
+    donutty.prototype.getOptionsFromTag = function() {
```
Comment (simeydotme) [inline src/donutty.js]:
```suggestion
```

File: src/donutty.js [hunk 4]
```diff
@@ -46,16 +58,24 @@
...
+
+        return JSON.parse(JSON.stringify(this.$wrapper.dataset));
```
Comment (simeydotme) [inline src/donutty.js]:
```suggestion
```

File: src/donutty.js [hunk 5]
```diff
@@ -46,16 +58,24 @@
...
+
+    };
```
Comment (simeydotme) [inline src/donutty.js]:
```suggestion
```

File: src/jquery.js [hunk 6]
```diff
@@ -6,14 +6,11 @@
 
         $( function() {
 
-            $.fn.donutty = function( options ) {
+            $.fn.donutty = function() {
```
Comment (simeydotme) [inline src/jquery.js]:
```suggestion
            $.fn.donutty = function( options ) {
```
bring back imperative options

File: src/jquery.js [hunk 7]
```diff
@@ -6,14 +6,11 @@
...
 
-                return $( this ).each( function( k, el ) {
+                return $( this ).each( function() {
 
-                    var $el = $( el ),
-                        instance = new Donutty( el, $.extend( {}, $el.data(), options ) );
-
-                    $el.data( "donutty", instance );
+                    new Donutty( this, null );
```
Comment (simeydotme) [inline src/jquery.js]:
```suggestion
                    new Donutty( this, options );
```

File: src/vanilla.js [hunk 8]
```diff
@@ -9,10 +9,7 @@
 
             Array.prototype.forEach.call( $donuts , function( $el ) {
 
-                var options = JSON.parse( JSON.stringify( $el.dataset ) ),
-                    instance = new Donutty( $el, options );
-
-                $el.dataset.donutty = instance;
+                new Donutty( $el, null );
```
Comment (simeydotme) [inline src/vanilla.js]:
```suggestion
                new Donutty( $el );
```

**Comment by simeydotme** [comment]
oh my god. 

This review process/interface is a disaster...
... I hope you're able to read/understand what I've changed, @i1group ... apologies for how fugly that is!!

----

*disclaimer:* I never tried that before, and I probably wont again, lol.

**Comment by i1group** [comment]
Fixed, check please.
I think it's better to keep `getOptionsFromTag()` method because it allows easily to grab settings from some donut and apply them to another instance.

**Comment by simeydotme** [comment]
## merged.

> I think it's better to keep getOptionsFromTag() method because it allows easily to grab settings from some donut and apply them to another instance.  

^ Ok, that's fair, I thought it would fail during initialisation but I tested and seems fine. 🙇 

> BTW would multiple calls of new Donutty('.donut') produce multiple donuts in one container?  

^ I guess so, but that's at the implementer's discretion. I would, however, like to add a simple `destroy()` method in future.

**Pull request merged**